File: agentguard/sdk/decorators.py
"""SDK decorators (BLUEPRINT-7 section 6.5.1).

Decorating a function records one ``TraceStep`` per call into the active
trace, if any.  The decorators support both sync and async callables and
accept either bare-decorator (``@tool``) or factory (``@tool(name="x")``)
syntax.

Default capture: arguments and return values are JSON-serialized with a
size cap (``AGENTGUARD_MAX_CAPTURE_BYTES``, default 4096). Sensitive
field names registered via ``scrub()`` are replaced with ``"***"``.
"""

from __future__ import annotations

import asyncio
import functools
import inspect
import json
import os
import time
from collections.abc import Callable, Iterable
from typing import Any

from agentguard.schemas.trace import TraceStep, TraceStepType
from agentguard.sdk.context import record_step

DEFAULT_MAX_CAPTURE_BYTES = 4096

_scrub_keys: set[str] = {
    "password",
    "secret",
    "api_key",
    "apikey",
    "token",
    "authorization",
    "openai_api_key",
    "anthropic_api_key",
}


def scrub(*keys: str) -> None:
    """Register additional argument / output keys to redact when captured.

    Matches against dict keys at any depth (case-insensitive). Already-included
    keys: password, secret, api_key, token, authorization, openai_api_key,
    anthropic_api_key.
    """
    for k in keys:
        _scrub_keys.add(k.lower())


def _max_bytes() -> int:
    try:
        return int(os.getenv("AGENTGUARD_MAX_CAPTURE_BYTES", str(DEFAULT_MAX_CAPTURE_BYTES)))
    except ValueError:
        return DEFAULT_MAX_CAPTURE_BYTES


def _redact(value: Any) -> tuple[Any, list[str]]:
    """Return (redacted_value, list_of_scrubbed_field_paths)."""
    redactions: list[str] = []

    def walk(node: Any, path: str) -> Any:
        if isinstance(node, dict):
            out: dict[str, Any] = {}
            for k, v in node.items():
                key_lower = str(k).lower()
                child_path = f"{path}.{k}" if path else str(k)
                if key_lower in _scrub_keys:
                    out[k] = "***"
                    redactions.append(child_path)
                else:
                    out[k] = walk(v, child_path)
            return out
        if isinstance(node, (list, tuple)):
            return [walk(v, f"{path}[{i}]") for i, v in enumerate(node)]
        return node

    return walk(value, ""), redactions


def _jsonable(value: Any) -> Any:
    """Convert value to a JSON-serializable form, truncating if too large."""
    try:
        text = json.dumps(value, default=str)
    except (TypeError, ValueError):
        text = repr(value)
    cap = _max_bytes()
    if len(text) > cap:
        text = text[:cap] + f"...<truncated {len(text) - cap} bytes>"
        return text
    try:
        return json.loads(text)
    except (TypeError, ValueError):
        return text


def _capture_args(fn: Callable, args: tuple, kwargs: dict) -> tuple[dict, list[str]]:
    """Bind args/kwargs to parameter names; return (captured_dict, redactions)."""
    try:
        sig = inspect.signature(fn)
        bound = sig.bind_partial(*args, **kwargs)
        captured: dict[str, Any] = dict(bound.arguments)
    except (TypeError, ValueError):
        captured = {"args": list(args), "kwargs": kwargs}
    redacted, redactions = _redact(captured)
    return _jsonable(redacted), redactions


def _build_step(
    *,
    kind: TraceStepType,
    name: str,
    captured_input: Any,
    output: Any,
    started_at: float,
    capture_input: bool,
    capture_output: bool,
    metadata: dict[str, Any] | None,
    input_redactions: list[str],
) -> TraceStep:
    output_value: Any = None
    output_redactions: list[str] = []
    if capture_output:
        redacted_output, output_redactions = _redact(output)
        output_value = _jsonable(redacted_output)
    return TraceStep(
        type=kind,
        name=name,
        input=captured_input if capture_input else None,
        output=output_value,
        metadata=metadata or {},
        latency_ms=int((time.time() - started_at) * 1000),
        source="sdk",
        redactions=input_redactions + output_redactions,
    )


def _wrap(
    fn: Callable,
    *,
    kind: TraceStepType,
    label: str,
    capture_input: bool,
    capture_output: bool,
    metadata: dict[str, Any] | None,
):
    is_coro = asyncio.iscoroutinefunction(fn)

    @functools.wraps(fn)
    def sync_wrapper(*args: Any, **kwargs: Any) -> Any:
        captured, redactions = (
            _capture_args(fn, args, kwargs) if capture_input else (None, [])
        )
        started = time.time()
        try:
            output = fn(*args, **kwargs)
        except Exception as exc:
            record_step(
                TraceStep(
                    type="error",
                    name=label,
                    input=captured,
                    output={"error_type": type(exc).__name__, "message": str(exc)},
                    metadata=metadata or {},
                    latency_ms=int((time.time() - started) * 1000),
                    source="sdk",
                    redactions=redactions,
                )
            )
            raise
        record_step(
            _build_step(
                kind=kind,
                name=label,
                captured_input=captured,
                output=output,
                started_at=started,
                capture_input=capture_input,
                capture_output=capture_output,
                metadata=metadata,
                input_redactions=redactions,
            )
        )
        return output

    @functools.wraps(fn)
    async def async_wrapper(*args: Any, **kwargs: Any) -> Any:
        captured, redactions = (
            _capture_args(fn, args, kwargs) if capture_input else (None, [])
        )
        started = time.time()
        try:
            output = await fn(*args, **kwargs)
        except Exception as exc:
            record_step(
                TraceStep(
                    type="error",
                    name=label,
                    input=captured,
                    output={"error_type": type(exc).__name__, "message": str(exc)},
                    metadata=metadata or {},
                    latency_ms=int((time.time() - started) * 1000),
                    source="sdk",
                    redactions=redactions,
                )
            )
            raise
        record_step(
            _build_step(
                kind=kind,
                name=label,
                captured_input=captured,
                output=output,
                started_at=started,
                capture_input=capture_input,
                capture_output=capture_output,
                metadata=metadata,
                input_redactions=redactions,
            )
        )
        return output

    return async_wrapper if is_coro else sync_wrapper


_RESERVED = {"name", "capture_input", "capture_output", "metadata"}


def _factory(default_kind: TraceStepType):
    def deco(
        *dargs: Any,
        name: str | None = None,
        capture_input: bool = True,
        capture_output: bool = True,
        metadata: dict[str, Any] | None = None,
        **extra: Any,
    ):
        # Any kwarg that is not one of the reserved decorator-control kwargs is
        # folded into the step metadata. This lets users write
        # ``@llm_call(model="gpt-4.1-mini", temperature=0)`` ergonomically.
        merged_metadata = {**(metadata or {}), **extra}

        # Bare decorator usage: @tool def foo(): ...
        if dargs and callable(dargs[0]) and not isinstance(dargs[0], type):
            fn = dargs[0]
            return _wrap(
                fn,
                kind=default_kind,
                label=name or fn.__name__,
                capture_input=capture_input,
                capture_output=capture_output,
                metadata=merged_metadata,
            )

        # Parametrized usage: @tool(name="x") def foo(): ...
        def wrap(fn: Callable) -> Callable:
            return _wrap(
                fn,
                kind=default_kind,
                label=name or fn.__name__,
                capture_input=capture_input,
                capture_output=capture_output,
                metadata=merged_metadata,
            )

        return wrap

    return deco


tool = _factory("tool_call")
llm_call = _factory("llm_call")
retrieval = _factory("retrieval")
guardrail = _factory("guardrail")
traceable = _factory("tool_call")  # generic alias for any user-defined step


__all__: Iterable[str] = [
    "tool",
    "llm_call",
    "retrieval",
    "guardrail",
    "traceable",
    "scrub",
]

File: packages/agentguard-js/src/decorators.ts
/**
 * Function-style "decorators" that record a TraceStep per call into the
 * active trace. Sync and async callables both supported.
 */

import { recordStep } from "./context.js";
import type { TraceStep } from "./schema.js";

type StepKind = TraceStep["type"];
type AnyFn<TArgs extends unknown[], TReturn> = (...args: TArgs) => TReturn | Promise<TReturn>;
type MaybeFn<TArgs extends unknown[], TReturn> = AnyFn<TArgs, TReturn> | DecoratorOptions;

export type DecoratorOptions = {
  name?: string;
  captureInput?: boolean;
  captureOutput?: boolean;
  metadata?: Record<string, unknown>;
};

const DEFAULT_MAX_BYTES = 4096;

const scrubKeys: Set<string> = new Set([
  "password",
  "secret",
  "api_key",
  "apikey",
  "token",
  "authorization",
  "openai_api_key",
  "anthropic_api_key",
]);

export function scrub(...keys: string[]): void {
  for (const k of keys) scrubKeys.add(k.toLowerCase());
}

function maxBytes(): number {
  const env = typeof process !== "undefined" ? process.env?.AGENTGUARD_MAX_CAPTURE_BYTES : undefined;
  if (!env) return DEFAULT_MAX_BYTES;
  const n = Number.parseInt(env, 10);
  return Number.isFinite(n) && n > 0 ? n : DEFAULT_MAX_BYTES;
}

function redactWalk(value: unknown, path: string, redactions: string[]): unknown {
  if (value && typeof value === "object" && !Array.isArray(value)) {
    const out: Record<string, unknown> = {};
    for (const [k, v] of Object.entries(value as Record<string, unknown>)) {
      const childPath = path ? `${path}.${k}` : k;
      if (scrubKeys.has(k.toLowerCase())) {
        out[k] = "***";
        redactions.push(childPath);
      } else {
        out[k] = redactWalk(v, childPath, redactions);
      }
    }
    return out;
  }
  if (Array.isArray(value)) {
    return value.map((v, i) => redactWalk(v, `${path}[${i}]`, redactions));
  }
  return value;
}

function jsonable(value: unknown): unknown {
  let serialized: string;
  try {
    serialized = JSON.stringify(value, (_k, v) => (typeof v === "bigint" ? v.toString() : v));
  } catch {
    return String(value);
  }
  if (serialized === undefined) return null;
  const cap = maxBytes();
  if (serialized.length > cap) {
    return serialized.slice(0, cap) + `...<truncated ${serialized.length - cap} bytes>`;
  }
  try {
    return JSON.parse(serialized);
  } catch {
    return serialized;
  }
}

function captureArgs(args: unknown[], paramNames: string[] | null): {
  captured: unknown;
  redactions: string[];
} {
  const redactions: string[] = [];
  let captured: unknown;
  if (paramNames && paramNames.length === args.length) {
    captured = Object.fromEntries(paramNames.map((n, i) => [n, args[i]]));
  } else {
    captured = { args };
  }
  const redacted = redactWalk(captured, "", redactions);
  return { captured: jsonable(redacted), redactions };
}

function inferParamNames(fn: Function): string[] | null {
  // Best-effort: parse the function source; not always available in minified
  // environments, in which case we fall back to the args-array shape.
  const src = fn.toString();
  const match =
    /function[^(]*\(([^)]*)\)/.exec(src) ??
    /^\s*(?:async\s*)?\(([^)]*)\)\s*=>/.exec(src) ??
    /^\s*(?:async\s*)?([^()\s,]+)\s*=>/.exec(src);
  if (!match || !match[1]) return null;
  return match[1]
    .split(",")
    .map((s) => s.trim().split(/[ :=]/)[0]!)
    .filter(Boolean);
}

function build<TArgs extends unknown[], TReturn>(
  kind: StepKind,
  fn: AnyFn<TArgs, TReturn>,
  options: DecoratorOptions = {},
): AnyFn<TArgs, TReturn> {
  const label = options.name ?? fn.name ?? "anonymous";
  const captureInput = options.captureInput ?? true;
  const captureOutput = options.captureOutput ?? true;
  const metadata = options.metadata ?? {};
  const paramNames = inferParamNames(fn as unknown as Function);

  return ((...args: TArgs): TReturn | Promise<TReturn> => {
    const start = Date.now();
    const captured = captureInput ? captureArgs(args as unknown[], paramNames) : { captured: null, redactions: [] };

    const finishOk = (output: unknown): void => {
      let outputValue: unknown = null;
      const outputRedactions: string[] = [];
      if (captureOutput) {
        const redacted = redactWalk(output, "", outputRedactions);
        outputValue = jsonable(redacted);
      }
      recordStep({
        type: kind,
        name: label,
        input: (captureInput ? captured.captured : undefined) as TraceStep["input"],
        output: outputValue as TraceStep["output"],
        metadata,
        latency_ms: Date.now() - start,
        source: "sdk",
        redactions: [...captured.redactions, ...outputRedactions],
      });
    };

    const finishErr = (err: unknown): void => {
      recordStep({
        type: "error",
        name: label,
        input: (captureInput ? captured.captured : undefined) as TraceStep["input"],
        output: {
          error_type: (err as Error)?.name ?? "Error",
          message: (err as Error)?.message ?? String(err),
        } as TraceStep["output"],
        metadata,
        latency_ms: Date.now() - start,
        source: "sdk",
        redactions: captured.redactions,
      });
    };

    let result: TReturn | Promise<TReturn>;
    try {
      result = fn(...args);
    } catch (err) {
      finishErr(err);
      throw err;
    }

    if (result && typeof (result as Promise<TReturn>).then === "function") {
      return (result as Promise<TReturn>).then(
        (value) => {
          finishOk(value);
          return value;
        },
        (err: unknown) => {
          finishErr(err);
          throw err;
        },
      );
    }

    finishOk(result);
    return result;
  }) as AnyFn<TArgs, TReturn>;
}

export function tool<TArgs extends unknown[], TReturn>(
  fnOrName: AnyFn<TArgs, TReturn> | string,
  options?: MaybeFn<TArgs, TReturn>,
): AnyFn<TArgs, TReturn> | ((fn: AnyFn<TArgs, TReturn>) => AnyFn<TArgs, TReturn>) {
  if (typeof fnOrName === "function") {
    return build("tool_call", fnOrName, options as DecoratorOptions | undefined);
  }
  if (typeof options === "function") {
    return build("tool_call", options, { name: fnOrName });
  }
  return (fn: AnyFn<TArgs, TReturn>) => build("tool_call", fn, { ...options, name: fnOrName });
}

export function llmCall<TArgs extends unknown[], TReturn>(
  fnOrName: AnyFn<TArgs, TReturn> | string,
  options?: MaybeFn<TArgs, TReturn>,
): AnyFn<TArgs, TReturn> | ((fn: AnyFn<TArgs, TReturn>) => AnyFn<TArgs, TReturn>) {
  if (typeof fnOrName === "function") {
    return build("llm_call", fnOrName, options as DecoratorOptions | undefined);
  }
  if (typeof options === "function") {
    return build("llm_call", options, { name: fnOrName });
  }
  return (fn: AnyFn<TArgs, TReturn>) => build("llm_call", fn, { ...options, name: fnOrName });
}

export function retrieval<TArgs extends unknown[], TReturn>(
  fnOrName: AnyFn<TArgs, TReturn> | string,
  options?: MaybeFn<TArgs, TReturn>,
): AnyFn<TArgs, TReturn> | ((fn: AnyFn<TArgs, TReturn>) => AnyFn<TArgs, TReturn>) {
  if (typeof fnOrName === "function") {
    return build("retrieval", fnOrName, options as DecoratorOptions | undefined);
  }
  if (typeof options === "function") {
    return build("retrieval", options, { name: fnOrName });
  }
  return (fn: AnyFn<TArgs, TReturn>) => build("retrieval", fn, { ...options, name: fnOrName });
}

export function guardrail<TArgs extends unknown[], TReturn>(
  fnOrName: AnyFn<TArgs, TReturn> | string,
  options?: MaybeFn<TArgs, TReturn>,
): AnyFn<TArgs, TReturn> | ((fn: AnyFn<TArgs, TReturn>) => AnyFn<TArgs, TReturn>) {
  if (typeof fnOrName === "function") {
    return build("guardrail", fnOrName, options as DecoratorOptions | undefined);
  }
  if (typeof options === "function") {
    return build("guardrail", options, { name: fnOrName });
  }
  return (fn: AnyFn<TArgs, TReturn>) => build("guardrail", fn, { ...options, name: fnOrName });
}

export function traceable<TArgs extends unknown[], TReturn>(
  fnOrName: AnyFn<TArgs, TReturn> | string,
  options?: MaybeFn<TArgs, TReturn>,
): AnyFn<TArgs, TReturn> | ((fn: AnyFn<TArgs, TReturn>) => AnyFn<TArgs, TReturn>) {
  return tool<TArgs, TReturn>(fnOrName, options);
}
